tool to add AnnotSV annotations to SV-VCF

ViktorHy opened the pull request:
<!--
Thanks for contributing to the CMD nextflow_wgs pipeline!

Please use the checklists below to document changes and performed tests.

Remember that this template doubles as our test/review documentation. 
-->
# Description and reviewer info

What is changed? How does this update improve the pipeline? (For reviewers) How to test it?


## Type of change
<!--
    Major change counts as a change that breaks backward compatibility
    Minor change is a substantial change that requires testing before deployment
    Patch is a minor change like a bug fix, code comment/style fix, etc.
-->
- [ ] Documentation
- [ ] Patch
- [ ] Minor change
- [x] Major change 

# Checklist

- [x] Self-review of my code
- [x] Update the CHANGELOG
- [ ] Tag the latest commit (vX.Y.Z format)

<!--
    Select a checklist below based on selection under # Type of change
    and delete the sections that do not apply to this PR:
-->

## Documentation
- [ ] At least one other person has reviewed my changes (not required for trivial changes)

## Patch
- [ ] Stub run completes without errors or new warnings
- [ ] At least one other person has reviewed and approved my code (not required for trivial changes)

## Major / Minor change
- [x] Stub run completes without errors or new warnings
- [x] GIAB single (wgs) finishes and differences to current master branch have been investigated (using [PipeEval](https://github.com/Clinical-Genomics-Lund/PipeEval))
- [x] GIAB trio (wgs) finishes and differences to current master branch have been investigated
- [x] Seracare (onco) sample finishes and differences to current master branch have been investigated
- [x] Output from processes with altered code have been inspected (i.e. in the work folder: .command.sh, .command.log, .command.err and result files)
- [x] All three samples can be loaded into Scout
- [x] At least one other person has reviewed and approved my code
- [x] I have made corresponding changes to the documentation (software versions, etc.)

# Test/review documentation

## Review performed by

- [x] Alexander
- [x] Jakob
- [ ] Paul
- [ ] Ryan
- [ ] Viktor

(Add if missing)

## Testing performed by

- [ ] Alexander
- [ ] Jakob
- [ ] Paul
- [ ] Ryan
- [x] Viktor

Jakob37 commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
+
+from pysam import VariantFile
+from pprint import pprint
```
Never used?

ViktorHy replied:
well it was used intensely during development =)

Jakob37 replied:
I suspected that 😄

Jakob37 commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+    
+    vcf_header = join_headers(vcf_object.header,annotsv_headers)
+    vcf_out = VariantFile(args.out_vcf, "w", header=vcf_header)
+    vcf_out.close()
```
Or is it maybe because you want to use the built in header writing, but do custom record writing?

Might be worth leaving a comment here to clarify if so

ViktorHy replied:
yes, cannot use the build in function to write the records. Will clarify

Jakob37 commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+        simple_id = f"{var.chrom}_{var.pos}_{svtype}"
+        if simple_id in annotsv_variants:
+            var_dict['INFO'] = add_annotsv_annotations(var_dict['INFO'],annotsv_variants[simple_id]['INFO'])
+        vcf_str = cmdvcf.print_vcf_record(var_dict,vcf_object.header)
```
Confusing method signature. `print` makes me think the function will write to STDOUT, but it returns a string.

I would prefer `cmdvcf.get_vcf_record_string(...)` or something similar

ViktorHy replied:
yes, it started out as printing, but that was a bit weird. Can rename the function

Jakob37 commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+    """
+    read VCF from annotsv
+    """
+    vcf_in = VariantFile(file_path, "r")
```
Do you need to close it again?

I think you can do a scoped opening here as well. Then you don't need to worry about remembering closing it.

```
with VariantFile(file_path, "r") as vcf_in:
  ...
```

Jakob37 commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+    annotsv_headers = get_relevant_headers(vcf_in)
+    return annotsv_variants,annotsv_headers
+
+def get_relevant_headers(vcf_in):
```
Can you make this function name more self-explanatory? "relevant" is a bit vague

ViktorHy replied:
remove_non_relevant_headers() *giggles*

Jakob37 commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+    """
+    headers = []
+    for key, record in vcf_in.header.info.items():
+        if 'ample' in key or 'cohort' in key:
```
No need to use a pass here, reverse it:

```
if key not in {'ample', 'cohort'}:
```

ViktorHy commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+    for var in vcf_object.fetch():
+        # to fix END flag disapearing from pysam object
+        # pysam eats up END from original, don't know why. (PERL vfc2.pm <3)
+        var_dict = cmdvcf.parse_variant(var,vcf_object.header)
```
dont understand, the code is already in cmdvcf? or you mean adding the END-flag in cmdvcf? Could do that, but would not create a specific for sv-varants cause that would be very redundant and I think you can have END-flags even for smaller indels

Jakob37 commented:
Looks very nice! Easy to follow and overall clear and well structured.

A bunch of smaller comments.

I found two major issues:

1. You repeatedly open the file when writing. You just need to open it once (or maybe two, if you want to separate out the header writing).
2. You mutate references to pass in objects in several functions. This makes no sense in the functions here, if thinking about what they actually do, and it is a dangerous habit which can cause severe headache (I have had one or two of those)

ViktorHy commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+        if simple_id in annotsv_variants:
+            var_dict['INFO'] = add_annotsv_annotations(var_dict['INFO'],annotsv_variants[simple_id]['INFO'])
+        vcf_str = cmdvcf.print_vcf_record(var_dict,vcf_object.header)
+        with open(args.out_vcf, "a") as vcf_out:
```
yeah, oops

ViktorHy commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,120 @@
+    """
+    for key in annotation:
+        if key not in original_info:
+            original_info[key] = annotation[key]
```
you are so easily scared =), but I was actually thinking about it.

Jakob37 replied:
Agreed, I have a pretty careful personality. But this is a quite significant one I would say. I have seen this pattern explode many times into horrible bugs (both for myself and others) 💣

ViktorHy commented:
@Jakob37 two more python scripts. I have not run black formatting on any so bare that in mind. Can you check whether I seem to fullfill the functions of prescore_sv.pl with these three scripts, and ofc do look at the code python code itself

Jakob37 commented:
Let's discuss type hintings a bit more at some point. They take an extra effort, but helps quickly capture many silly errors.

Here watching `add_annotsv.py` with "standard" type checking mode. 

Settings

![type_checking_mode2](https://github.com/user-attachments/assets/79099862-14e8-4265-8aaf-5b8b7f380bba)

Type hints

![typing](https://github.com/user-attachments/assets/2635bb2e-3cfe-47c9-b27b-b48509b6635d)

Jakob37 commented:
From CLI:

![mypy](https://github.com/user-attachments/assets/b1747627-e4ac-4af7-be15-be8c01995361)

Jakob37 commented on bin/add_geneticmodels_to_svvcf.py:
```diff
@@ -0,0 +1,146 @@
+        "String",
+        "If variant follows geneticmodel and is a homo- or hemi-zygous deletion"
+    )
+    vcf_out = VariantFile(args.out_vcf, "w", header=vcf_object.header)
```
To make the `main` function neater, you could put this into an `output_vcf` function

Jakob37 commented on bin/geneticmodels.py:
```diff
@@ -0,0 +1,291 @@
+# Proband Sex   Proband Genotype    Xchrom(bool)    Father Genotype Father Affections   Mother Genotype Mother Affection
```
Hmmm. Is this file coming from a specific source? Would be good with a reference to better understand it.

Also, is there a point with the tab delimited header?

ViktorHy replied:
Only because I like it. I handcrafted this file, you will like it

alkc replied:
Please cut all 'NA' fields. The script reading from this already returns 'NA' for codes that do not have a GM defined.

Jakob37 commented on bin/add_geneticmodels_to_svvcf.py:
```diff
@@ -0,0 +1,146 @@
+            vcf_out.write(vcf_str)
+
+
+def genotype_sum_per_individual(gt:dict):
```
I will not comment on all of them, but in my mind it would be good to try doing this typing for all functions.

Let's discuss though. I am in the habit of doing this, and not sure how much extra work it means for you.

ViktorHy replied:
I mean it's a broader discussion, for me it's 1% that takes me 20% of time that I dont feel like is worth it. I need to move forward with other projects

Jakob37 commented on bin/add_geneticmodels_to_svvcf.py:
```diff
@@ -0,0 +1,146 @@
+    """
+    ped, individuals = read_ped(args.pedigree_file)
+    if len(ped.keys()) < 2:
+        exit("Cannot add genetic models to single samples")
```
`sys.exit` I think (`exit` does not seem to be intended for production)

https://www.geeksforgeeks.org/python-exit-commands-quit-exit-sys-exit-and-os-_exit/

Jakob37 commented on bin/add_geneticmodels_to_svvcf.py:
```diff
@@ -0,0 +1,146 @@
+    """
+    sample_gt_sum = genotype_sum_per_individual(var_dict['GT'])
+    chrom = var_dict['CHROM']
+    x = 0
```
Is this to indicate that "chromosome is X"?

Could you do this instead:

```
is_x = False
if chrom == "X":
  is_x = True
```

Jakob37 commented on bin/add_geneticmodels_to_svvcf.py:
```diff
@@ -0,0 +1,146 @@
+def is_del_hemi_homo(proband_gt,proband_sex,x):
+    """
+    checks if deletion is complete loss of genetic material
+    Boys on X special, ladies on X maybe not needed to check
```
Could you give a mention of what the `proband_sex` values mean

Jakob37 replied:
Or maybe even better. Why not put constants at the top of the module

```
SEX_MALE = 1
SEX_FEMALE = 2
```

and then

```
if proband_sex == SEX_MALE and proband_gt < 0:
```

Jakob37 commented on bin/add_geneticmodels_to_svvcf.py:
```diff
@@ -0,0 +1,146 @@
+        for line in f:
+            fields = line.strip().split('\t')
+            ind = {
+                "FATHER": fields[2],
```
For a later reader, it is not trivial to understand what is in all the Peddy-fields

Although a bit more verbose, I think constants would help readability here

```
father = fields[2]
mother = fields[3]
sex = fields[4]
phenotype = fields[5]
```

ViktorHy replied:
looking back to this, I did actually have to remind myself about the pedigreefile format. I'll add this =)

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+        "String",
+        "big manta events not supported by gatk/cnvnator"
+    )
+    vcf_out = VariantFile(args.out_vcf, "w", header=vcf_object.header)
```
What about putting this into an `output_vcf` function to keep the main function clean

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+
+MANTA_SIZE_FOR_PENALTY = 100000
+
+def main(args: object):
```
`argparse.Namespace`

or, my preference:

```
def main(input_vcf: str, out_vcf: str):
```

Called as such

```
main(args.input_vcf, args.out_vcf)
```

Then typing and potential unit testing becomes easier

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+                vcf_str = cmdvcf.vcf_string(var_dict,vcf_object.header)
+            vcf_out.write(vcf_str)
+
+def variant_called_by(varcallers: str):
```
```suggestion
def variant_called_by(varcallers: str) -> List[str]:
```

`List` with import from `typing` library for up to Python 3.8.

`list` without import afterwards.

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+        varcallers_list.append('tiddit')
+    return varcallers_list
+
+def modify_gatk(var_dict: dict):
```
The `modify_gatk` name does not seem correspond to the docstring and what the function seem to be doing

Jakob37 replied:
`get_average_gt_quality` ?

ViktorHy replied:
same as the manta part, debating names, thinking of future use of the function.

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+        if genotype_qual is not None and genotype_qual != '.':
+            genotype_qual_sum = genotype_qual_sum + int(genotype_qual)
+            number_of_qual_values = number_of_qual_values + 1
+    mean_qual_val = int(genotype_qual_sum/number_of_qual_values)
```
Could you end up with `number_of_qual_values == 0` here? If so, this would crash the function.

ViktorHy replied:
Hmm never happened with the perl-script, but could perhaps take that into account

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+            var_dict = cmdvcf.parse_variant(var,vcf_object.header)
+            varcall_set = var_dict['INFO'].get('set',None)
+            if varcall_set is None:
+                exit("need a VCF merged with SVDB")
```
```suggestion
                sys.exit("need a VCF merged with SVDB")
```

ViktorHy replied:
pasty boy coming in hot

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+        
+
+def parse_arguments():
+    parser = argparse.ArgumentParser(description="Adds genetic models to SV-variants, needs a matching pedigree file")
```
Update the description for this script (this looks like a copy paste from the other script)

Jakob37 commented on bin/sv_varcall_penalties.py:
```diff
@@ -0,0 +1,112 @@
+    mean_qual_val = int(genotype_qual_sum/number_of_qual_values)
+    return mean_qual_val
+
+def modify_manta(var_dict: dict):
```
`get_manta_penalty` ?

No modification is done within this function (it is done in a broader scope, beyond what needs to be considered to use this function)

ViktorHy replied:
Was debating this in my head, think I ultimately wanted to give us options to further do manta filtering that required some more fancy calculcations

ViktorHy commented:
Update 250228. Introduced all three python scripts into nextflow code. Removed unused add_omim.pl. Added a new process to annotate dbvar accession nr, clinvar accession number and dbvar/clinvar status (pathogenic/likely pathogenic)

ViktorHy commented:
TODO: Add gnomadSV via vep, and update rank model field names and should be good for testing in pipeeval

alkc commented on bin/reference_tools/banta_gnomad_sv.py:
nice. good to have this around.

alkc reviewed (changes requested):
main comments are the omim stuff in the rank model and the new gnomad svAF annotation

ViktorHy commented on rank_models/tmp/sv_rank_model_panel_v1.1.ini:
```diff
@@ -0,0 +1,522 @@
+  info_key = OMIM_morbid
+  description = reported morbid genes
+  
+  [[Found]]
+    score = 3
+    priority = 0
+    string = 'yes'
+    
+  [[not_reported]]
+    score = 3
```
yes, this is actually kind of silly. Should be documented somewhere. Since this is a panel, all genes we included in the design is of interest and as such, all should get points for that.

alkc replied:
gotcha

ViktorHy commented on main.nf:
```diff
@@ -3620,6 +3676,7 @@ process vep_sv {
 			--dir_plugins $params.VEP_PLUGINS \\
 			--max_sv_size $params.VEP_MAX_SV_SIZE \\
 			--distance $params.VEP_TRANSCRIPT_DISTANCE \\
+			--custom file=$params.GNOMAD_SV,short_name=gnomad,fields=AF%FREQ_HOMALT%SVTYPE,format=vcf,reciprocal=1,overlap_cutoff=70,same_type=1 \\
```
checked in scout. It works!

alkc replied:
when i compare yr onco pipeeval sample with an older onco run, a bunch of DUPs have lost their AF annotation. is this expected?

ViktorHy replied:
I think mostly it is because I require reciprocal overlap of 70% much like what we do for all other allele frequencies. AnnotSV, where the previous annotation came from used a non-reciprocal overlap. I tried this for gnomadSV in vep but just ended up nuking all our important variants from orbit

ViktorHy replied:
Also, DUPs are a bit special, since gnomad does not differentiate between TDUPs and DUPs in their data.

alkc commented on main.nf:
ok!

Jakob37 commented on bin/reference_tools/banta_gnomad_sv.py:
Can we move this to resources

ViktorHy commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,170 @@
+    """
+    headers = []
+    for key, record in vcf_in.header.info.items():
+        if key not in ['ample', 'cohort']:
```
I cannot even make sense of this any longer. Looking at the VCF I used for testing the key should never be just cohort or just sample/Sample, but contain that word. Seems like this should be reversed. if cohort in key.

ViktorHy replied:
or does not in work in both directions?

Jakob37 replied:
OK, you have a string "cohort_blabla" or something? And you want to skip those that contains that pattern ..

If you do

```
>>> "cohort" in "cohort_blabla"
True
```

But you want to check any of many strings isn't it

This might work, if what you are after?

```
>>> [target in "my_cohort_key" for target in ["cohort", "sample"]]
[True, False]
>>> any([target in "my_cohort_key" for target in ["cohort", "sample"]])
True
```

Jakob37 replied:
I still see `if key not in ['ample', 'cohort']:` which I suspect will never work. Have you pushed the changes?

> Maybe I should just remove the vcf-function.

Yes, if not used, rather remove than keeping it untested

Jakob37 commented on bin/add_annotsv.py:
```diff
@@ -0,0 +1,170 @@
+        annotsv_variants, annotsv_headers = read_annotsv_vcf(args.annotsv)
+    
+    vcf_header = join_headers(original_header,annotsv_headers)
+    vcf_out = VariantFile(args.out_vcf, "w", header=vcf_header)
```
OK, I think the same. A personal preference then would be to crash early and clearly:

```
if not args.out_vcf.endswith(".vcf"):
  print("Input file must end with .vcf")
  sys.exit(1)
```

But up to you!
